### PR TITLE
Adding configurable font colors.

### PR DESCRIPTION
**Pull request by CoryDCode**
This change is [Reviewable](https://reviewable.io/reviews/google/blockly-ios/336)

### CODE REVIEW

**Comment by CoryDCode** [comment]
+@vicng +@roboerikg 

Note: I'm leaving the default black for the colors because the blocks in the demo necessitate it. We can change those in a future CL.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336#-:-Kg0YZU8xT0l75rvfnlW:bm6mwws)*

File: Source/UI/View Controllers/DropdownOptionsViewController.swift [hunk 1]
```diff
@@ -67,6 +67,8 @@ open class DropdownOptionsViewController: UITableViewController {
       calculatePreferredContentSize()
     }
   }
+  /// The color of the font on each cell
+  open var textLabelColor: UIColor? = .black
```
Comment (RoboErikG) [inline Source/UI/View Controllers/DropdownOptionsViewController.swift]: Why doesn't this use color(for: LayoutConfig.FieldInputTextColor)?

**Comment by CoryDCode** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/UI/View Controllers/DropdownOptionsViewController.swift, line 71 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg0YXZMbQ9Qi7z787kD-r1-71:-Kg0aBUvopocmP-_HgWx:b-un3o2p) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/View Controllers/DropdownOptionsViewController.swift#L71)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Why doesn't this use color(for: LayoutConfig.FieldInputTextColor)?
</blockquote></details>

I'm not positive, but I'm guessing that it can't be initialized to that because the VC isn't associated with a layout until it's presented. None of the other elements of this class default that way, either, and I assumed that was the reason, but I'll have to ask @vicng .

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by vicng** [comment]
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Layout/LayoutConfig.swift, line 71 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CqLK6OCPVQA5T_WA:-Kg1CqLLMM7O9Tjx90mI:b2eopb9) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/Layout/LayoutConfig.swift#L71)):*
> ```Swift
> 
>   /// [`UIColor`] The default color for text in fields.
>   public static let FieldTextColor = LayoutConfig.newPropertyKey()
> ```

Rename to FieldLabelColor (I think it's only being used there?)

---

*[Source/Layout/LayoutConfig.swift, line 74 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CzjbE5hhc5De-cel:-Kg1CzjbE5hhc5De-cem:bfc6zrq) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/Layout/LayoutConfig.swift#L74)):*
> ```Swift
> 
>   /// [`UIColor`] The default color for text input in fields.
>   public static let FieldInputTextColor = LayoutConfig.newPropertyKey()
> ```

The name is confusing since we have a "FieldInput" class. Perhaps "FieldEditableTextColor"? (Although that's not a great name either since we have an "editable" property on fields...) Open to any ideas here.

---

*[Source/UI/View Controllers/DropdownOptionsViewController.swift, line 71 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg0YXZMbQ9Qi7z787kD-r1-71:-Kg1BESYKM128JG5vCxx:b-dgy7um) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/View Controllers/DropdownOptionsViewController.swift#L71)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

I'm not positive, but I'm guessing that it can't be initialized to that because the VC isn't associated with a layout until it's presented. None of the other elements of this class default that way, either, and I assumed that was the reason, but I'll have to ask @vicng .
</blockquote></details>

This is correct. DropdownOptionsViewController is decoupled from any specific Layout class, so it can be used by anything.

---

*[Source/UI/Views/FieldDropdownView.swift, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1BrN_ftkfrLsj9e7m:-Kg1BrN_ftkfrLsj9e7n:bg0ggbt) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldDropdownView.swift#L138)):*
> ```Swift
>       fieldDropdownLayout.config.popoverFont(for: LayoutConfig.GlobalFont)
>     viewController.textLabelColor =
>       fieldDropdownLayout.config.color(for: LayoutConfig.FieldInputTextColor)
> ```

This should use its own property. These are values being presented in a table view, which defaults to a white background (which could also be another property...it depends how configurable you want to get) 

---

*[Source/UI/Views/FieldVariableView.swift, line 135 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CJ0cmPLvIR8-60YL:-Kg1CJ0dq1U0MmgB2OsT:b-sdk4jt) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldVariableView.swift#L135)):*
> ```Swift
>       fieldVariableLayout.config.popoverFont(for: LayoutConfig.GlobalFont)
>     viewController.textLabelColor =
>       fieldVariableLayout.config.color(for: LayoutConfig.FieldInputTextColor)
> ```

Same thing here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by CoryDCode** [comment]
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Layout/LayoutConfig.swift, line 71 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CqLK6OCPVQA5T_WA:-Kg1OJr1yS7ZCCjJyfum:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/Layout/LayoutConfig.swift#L71)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Rename to FieldLabelColor (I think it's only being used there?)
</blockquote></details>

Done.

---

*[Source/Layout/LayoutConfig.swift, line 74 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CzjbE5hhc5De-cel:-Kg1OL3erY9qEN7QSv6J:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/Layout/LayoutConfig.swift#L74)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

The name is confusing since we have a "FieldInput" class. Perhaps "FieldEditableTextColor"? (Although that's not a great name either since we have an "editable" property on fields...) Open to any ideas here.
</blockquote></details>

Done.

---

*[Source/UI/Views/FieldDropdownView.swift, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1BrN_ftkfrLsj9e7m:-Kg1OM74E5B-TsE7asu2:b-96tp0h) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldDropdownView.swift#L138)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

This should use its own property. These are values being presented in a table view, which defaults to a white background (which could also be another property...it depends how configurable you want to get) 
</blockquote></details>

I disagree. I thought about that, but it's weird - why aren't the values on the popover the same color as the values on the dropdown? And yes, we could recolor all of those, but at some point, I think the answer is "extend/change View" if you want to change the default stuff - the dropdowns look wrong if they're not the same color as the inputs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by vicng** [comment]
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/UI/Views/FieldDropdownView.swift, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1BrN_ftkfrLsj9e7m:-Kg1SeUe12J_XlFvPHHO:b-h7ld2j) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldDropdownView.swift#L138)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

I disagree. I thought about that, but it's weird - why aren't the values on the popover the same color as the values on the dropdown? And yes, we could recolor all of those, but at some point, I think the answer is "extend/change View" if you want to change the default stuff - the dropdowns look wrong if they're not the same color as the inputs.
</blockquote></details>

I'd argue it's not an "editable text field" at this point. In any case, it's a minor inconvenience for users to add an extra line of code rather than to force them to always use the same color for both the dropdown and editable text fields.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by vicng** [comment]
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Layout/LayoutConfig.swift, line 70 at r2](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1T_r9yQmgtByui451:-Kg1T_r9yQmgtByui452:bols8kh) ([raw file](https://github.com/google/blockly-ios/blob/5a73ce0621aaabc24184f12c9e8f46bc52733049/Source/Layout/LayoutConfig.swift#L70)):*
> ```Swift
>   public static let FieldCheckboxSwitchTintColor = LayoutConfig.newPropertyKey()
> 
>   /// [`UIColor`] The default color for text in fields.
> ```

Update comment.

---

*[Source/Layout/LayoutConfig.swift, line 73 at r2](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1Tax1C5ehLGvwD6Bf:-Kg1Tax1C5ehLGvwD6Bg:b-tet5im) ([raw file](https://github.com/google/blockly-ios/blob/5a73ce0621aaabc24184f12c9e8f46bc52733049/Source/Layout/LayoutConfig.swift#L73)):*
> ```Swift
>   public static let FieldLabelTextColor = LayoutConfig.newPropertyKey()
> 
>   /// [`UIColor`] The default color for text input in fields.
> ```

Clarify comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by CoryDCode** [comment]
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Layout/LayoutConfig.swift, line 70 at r2](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1T_r9yQmgtByui451:-Kg1UWiF5h5mgcStcepg:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/5a73ce0621aaabc24184f12c9e8f46bc52733049/Source/Layout/LayoutConfig.swift#L70)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Update comment.
</blockquote></details>

Done.

---

*[Source/Layout/LayoutConfig.swift, line 73 at r2](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1Tax1C5ehLGvwD6Bf:-Kg1UXS3sHbQdArEs6L2:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/5a73ce0621aaabc24184f12c9e8f46bc52733049/Source/Layout/LayoutConfig.swift#L73)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Clarify comment.
</blockquote></details>

Done.

---

*[Source/UI/Views/FieldDropdownView.swift, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1BrN_ftkfrLsj9e7m:-Kg1X08DZfk2XDFXzUlz:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldDropdownView.swift#L138)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

I'd argue it's not an "editable text field" at this point. In any case, it's a minor inconvenience for users to add an extra line of code rather than to force them to always use the same color for both the dropdown and editable text fields.
</blockquote></details>

Done.

---

*[Source/UI/Views/FieldVariableView.swift, line 135 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1CJ0cmPLvIR8-60YL:-Kg1X-QALiVxi0SEFkhk:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldVariableView.swift#L135)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Same thing here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336)*

**Comment by vicng** [comment]
:lgtm:

---

Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/UI/Views/FieldDropdownView.swift, line 138 at r1](https://reviewable.io:443/reviews/google/blockly-ios/336#-Kg1BrN_ftkfrLsj9e7m:-Kg1ZWiI8AH_wkot1_Sq:b3unn2e) ([raw file](https://github.com/google/blockly-ios/blob/60e73b1e2afc1dfd6f9547caa60041a118b8e144/Source/UI/Views/FieldDropdownView.swift#L138)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

Done.
</blockquote></details>

My bad, you were right the first time. Lunch is on me next time :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336#-:-Kg1ZbgZPC-Qmfx6D40W:bnfp4nl)*

**Comment by RoboErikG** [comment]
:lgtm:

---

Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/336#-:-Kg1nyfxUAD-xsuIN_w9:bnfp4nl)*

**Pull request merged**